utils/minnesota.py

```diff
@@ -0,0 +1,423 @@
+            processed_dfs.append(processed_cand_con)
+        # noncandidate-recipient contribution data
+        noncandidate_df = pd.read_csv(filepaths_list[-2])
+        processed_noncandidate_con = self.preprocess_noncandidate_contribution(
+            noncandidate_df
+        )
+        processed_dfs.append(processed_noncandidate_con)
+        # expenditure data
+        expenditure_df = pd.read_csv(filepaths_list[-1])
+        processed_expenditure_df = self.preprocess_expenditure(expenditure_df)
+        processed_dfs.append(processed_expenditure_df)
+        combined_df = pd.concat(
+            processed_dfs,
+            ignore_index=True,
+        )
+
+        return [combined_df]
+
+    def clean(self, data: list[pd.DataFrame]) -> list[pd.DataFrame]:
+        """
+        Cleans MN data by converting dtyes to match database schema and dropping
+        non-classfiable rows (not represent minimal viable transactions)
+
+        Inputs:
+            data: a list of 1 DataFrame outputted from the preprocess method
+
+        Returns: a list of 1 cleaned MN DataFrame
+        """
+
+        df = data[0]
+        df["date"] = pd.to_datetime(df["date"])
+        df["year"] = df["date"].dt.year
+        df = df.drop(columns=["date"])
+        type_mapping = {
+            "state": "str",
+            "recipient_id": "str",
+            "donor_id": "str",
+            "recipient_first_name": "str",
+            "recipient_last_name": "str",
+            "recipient_full_name": "str",
+            "donor_first_name": "str",
+            "donor_last_name": "str",
+            "donor_full_name": "str",
+            "recipient_type": "str",
+            "donor_type": "str",
+            "purpose": "str",
+            "transaction_type": "str",
+            "year": "int64",
+            "amount": "float64",
+            "inkind_amount": "float64",
+        }
+        df = df.astype(type_mapping)
+        # non-classfiable rows of no transaction amount, no donor/recipient info
+        df = df[df["amount"] != 0]
+        df = df.dropna(subset=["recipient_id", "donor_id"], how="any")
+        df = df.drop(columns=["inkind_amount"])
+        df = df.reset_index(drop=True)
+
+        return [df]
+
+    def standardize(self, data: list[pd.DataFrame]) -> list[pd.DataFrame]:
+        """
+        Standardizes the dataframe into the necessary format for the schema
+
+        Maps entity/office types and column names as defined in schema, adjust
+        and add UUIDs as necessary, store base id and provided id in MNIDMap.csv
+
+        Inputs:
+            df: A list of 1 preprocessed and cleaned Dataframe outputted from
+            the clean method
+
+        Returns: A list of 1 standarized DataFrame matching database schema
+        """
+
+        df = data[0].copy()  # Create a copy to avoid modifying the original DataFrame
+        df["company"] = None  # MN dataset has no company information
```

Comment:
see slack for how to make this type of transformation more efficient throughout your code